src/ooni/ooni_test_impl.hpp

```diff
@@ -69,10 +70,40 @@ class OoniTestImpl : public mk::NetTest {
             });
     }
 
-    void geoip_lookup() {
+    void geoip_lookup(Callback<> cb) {
         probe_ip = "127.0.0.1";
         probe_asn = "AS0";
         probe_cc = "ZZ";
+        ip_lookup([=](Error err, std::string ip) {
+            if (!err) {
+                logger->info("probe ip: %s", ip.c_str());
+                if (options.get("save_real_probe_ip", false)) {
+                    logger->debug("saving user's real ip on user's request");
+                    probe_ip = ip;
+                }
+                std::string country_p = options.get("geoip_country_path",
+                                                    std::string{});
+                std::string asn_p = options.get("geoip_asn_path",
+                                                std::string{});
+                if (country_p == "" or asn_p == "") {
+                    logger->warn("geoip files not configured; skipping");
+                } else {
+                    ErrorOr<nlohmann::json> res = geoip(ip, country_p, asn_p);
+                    if (!!res) {
+                        logger->debug("GeoIP result: %s", res->dump().c_str());
+                        try {
+                            probe_asn = (*res)["asn"];
+                            logger->info("probe_asn: %s", probe_asn.c_str());
+                            probe_cc = (*res)["country_code"];
+                            logger->info("probe_cc: %s", probe_cc.c_str());
+                        } catch (std::domain_error &) {
+                            logger->warn("failed to access geoip result");
```

Comment:
Ah, you're right! I think it would be even better to let the exception unwind when there is an error here; since that is the usual path and errors here could only happen after some future, bad refactoring, I'd say we would notice for sure an exception while we may miss a warning. So, will remove the `try...catch` and instead add explanatory comment!